### ISSUE TITLE
package.json#imports/exports are not fully resolved for JS files

### ISSUE THREAD

**Issue by bmeck** [microsoft/TypeScript#57259]
### 🔎 Search Terms

imports, package.json, subpath imports, exports, aliasing

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about package.json imports

### ⏯ Playground Link

https://github.com/bmeck/typescript-issue-44848-demo

### 💻 Code

See repo link

### 🙁 Actual behavior

Paths are not resolved for JS files using package.json#exports/imports using moduleResolution "NodeNext" in either tsconfig or jsconfig.

### 🙂 Expected behavior

Modules should resolve in the same manner that Node does for JS files with corresponding TSConfig or JSConfig.

**Comment by Andarist** [comment]
Do you refer to the lack of go to the definition within your `self-import.js` file? Or do you mean that it doesn't work in `dist` directory?

If it's about first then I'd say that this works as intended. Your `tsconfig.json` doesn't include the `self-import.js` so its configuration is not applied to it.

If it's about the second then I'd say that this issue is the duplicate of a recent issue of mine: https://github.com/microsoft/TypeScript/issues/57064 😉

**Comment by Andarist** [comment]
By the way, you shouldn't need this `src/package.json` as long as your inputs match with outputs. TS is smart enough to figure out what source file it should map to based on the paths used within exports/imports as long as it can be directly mapped back using rootDir, outDir and similar configurations.

It doesn't work in this repo when I remove that `package.json` but when I set `"rootDir": "src"` then the navigation within the `src` goes to the respective source files (and not to output files in `dist`)

**Comment by bmeck** [comment]
@Andarist 
> By the way, you shouldn't need this src/package.json as long as your inputs match with outputs. TS is smart enough to figure out what source file it should map to based on the paths used within exports/imports as long as it can be directly mapped back using rootDir, outDir and similar configurations.

I'm quite confused with this paragraph, but the src/package.json seems simple enough as it has less entangling for tools outside TS itself so I'd lean towards not removing it either way.

> Or do you mean that it doesn't work in dist directory?

That is the main contention but in general normal JS resolution fails for even implicit project config via vscode etc. removing tsconfig entirely.

**Comment by Andarist** [comment]
> I'm quite confused with this paragraph,

I meant that when TS sees a imports like this:
```json
    "imports": {
        "#a/*": "./dist/*.js"
    },
```
and `tsconfig.json` like this:
```json
        "outDir": "dist",
        "rootDir": "src",
        "module": "NodeNext",
        "moduleResolution": "NodeNext",
```

then it's able to see that the output of `src/imported.ts` is `dist/imported.js`. It has to know that because it has to emit those files. In addition to that, it understands that `dist/imported.js` matches `package.json#imports` here and that it can map to *some* input file. So using the `rootDir` and `outDir` (and maybe some other things too) it can substitute the parts of the defined `package.json#imports` pattern to get the input file for this output file. The result is that it can understand that - when in `src` - it can go straight to the respective source file that matches the declared output pattern used by `package.json#imports`

> but the src/package.json seems simple enough as it has less entangling for tools outside TS itself so I'd lean towards not removing it either way.

Fair. I actually don't use the capability described above either. For maximum compatibility, I define `imports` that point to source files and my emitter replaces those `#`-based sources with the resolved relative paths. It works as long as all used conditions share the same interface.

> That is the main contention but in general normal JS resolution fails for even implicit project config via vscode etc. removing tsconfig entirely.

Yeah, that's a byproduct of `moduleResolution: node` still being the de facto default and `moduleResolution: node16` being an opt-in.

**Comment by RyanCavanaugh** [comment]
> mistargetted due to TS not rewriting them

TS doesn't rewrite paths during emit; this is intentional and on purpose. If the path you want isn't present in the output, change the input.

I can't otherwise tell from the repo what you think is wrong or needs to be changed. Please use the module resolution bug template.

**Comment by fatcerberus** [comment]
@RyanCavanaugh I'm not sure what you're quoting

**Comment by RyanCavanaugh** [comment]
It's in the linked repo

**Comment by bmeck** [comment]
@RyanCavanaugh 
> TS doesn't rewrite paths during emit; this is intentional and on purpose

Yes, the issue is not requesting any sort of output change or anysuch. The repo was mad as an example for #44848 and exposed some oddities.

Even changing the repo to have a jsconfig / removing tsconfig/src the paths still don't seem to resolve:

<img width="465" alt="Screenshot 2024-02-02 at 8 57 55 AM" src="https://github.com/microsoft/TypeScript/assets/234659/0aef4986-b039-470d-ad52-c0c83cbf7fc6">
<img width="525" alt="Screenshot 2024-02-02 at 8 58 15 AM" src="https://github.com/microsoft/TypeScript/assets/234659/960e17bf-8ed5-4265-980b-ca68f29c6f5a">

I wonder if this might be a VSCode bug, but am somewhat skeptical given using tsconfig/jsconfig seem to always be unable to resolve in that `dist/` repo directory.

**Comment by fatcerberus** [comment]
I'm not sure that's even a supported use case, to have a `jsconfig.json` and `tsconfig.json` side-by-side in the same directory.

**Comment by bmeck** [comment]
Without both, you can't quite make everything work and various imports
don't resolve in any simple configuration that I can tell depending on how
you tweak things. This does allow goto definition to work in both src and
dist directories.
* Putting `dist` into the tsconfig doesn't do anything that I can tell to
make `dist/main.*` able to goto `dist/imported.*` via goto definition.
* Only including `src` in `tsconfig.json` doesn't allow resolution from
`dist/main.*` to `dist/imported.*` either.
* Without the jsconfig, `dist` goto definition doesn't work, but adding it
does work

After poking and double checking here, having both files seems buggy / not
always picked up? So maybe that is a bug?
